[All] Improve extlibs integration

guparan opened the pull request:
This is a (big) attempt to clean the dependencies management in SOFA.

## Better understand SOFA dependencies
Here is a summary: https://docs.google.com/spreadsheets/d/1dxP7Rsyl1oCr0lfTwQESeeZQTt315CIFih5QpJp5OZY/edit?usp=sharing

## Clean dependency management in CMakeLists
- SofaFramework
  - Moved to SofaHelper: Boost, OpenGL, GLEW, GTest, Eigen, JSON, FFMPEG_exec  
  - Removed (not needed by any SofaFramework module): TinyXML, ZLIB 
- SofaMisc
  - Added to SofaMiscTopology: ZLIB
  - Added to SofaMiscFem: Newmat
- SofaGeneral
  - Added to SofaDenseSolver: Newmat
  - Added to SofaGeneralLoader: ZLIB
- SofaGui
  - Moved to SofaGuiQt: Qt + lots of options about QtViewer, QGLViewer, VSync, ...
- Modularized modules (oh yeah)
  - SofaPreconditioner: removed Newmat, added SofaSparseSolver
  - SofaExporter: added ZLIB
- Plugins
  - SofaMiscCollision: added SofaDistanceGrid
  - SofaImplicitField: added SofaPython
  - Registration: removed SofaOpenglVisual
  - image: added ZLIB
  - Flexible: added SofaPython

## Rework all YYY_HAVE_XXX variables to reduce their scope

This is important for future global modularization of SOFA.

- Variables are now named upon their module. 

- Whenever you need to use the YYY_HAVE_XXX of another module, you surely should declare your own MYMODULE_HAVE_XXX instead.

- All those new YYY_HAVE_XXX variables are booleans, so they can be used in CMake with `if(YYY_HAVE_XXX)` and in C++ (if correcly propagated via a config.h.in file) with `#if YYY_HAVE_XX` or `if(YYY_HAVE_XXX)`.

- Add sofa_find_package and sofa_set_bool macros to auto create YYY_HAVE_XXX variables.

- Edit sofa_install_targets to call `configure_file()` on any *.h.in found in target sources (the sources that are given to `add_library()`).

:warning: Here is a quick tuto about [how to correctly add and propagate a dependency](https://gist.github.com/guparan/0b0ed00ad60b740f85b794cdff255229).

## Rework #1121 to avoid nested projects

- Use SOFA_NO_OPENGL to build an OpenGL-free version of SofaGuiQt



______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

guparan committed:
[CMake] CLEAN Sofa*Targets.cmake inclusions and SofaComponent* packages

guparan committed:
[All] CLEAN Boost::program_options usage

namespace alias "po" makes difficult to find where program_options is used

guparan committed:
[SofaTest_test] Disable project

SofaTest_test contains only:
- scenes/damping.py
- python_tests/python_test_test.py

If this scripts are needed, they should be in SofaPython_test.

guparan committed:
[gtest] CLEAN include directories

guparan committed:
[MiniFlowVR] CLEAN include directories

guparan committed:
[Newmat] CLEAN Config.cmake

guparan committed:
[TinyXML] CLEAN Config.cmake

guparan committed:
[SofaGTestMain] CLEAN CMakeLists.txt

guparan committed:
[All] Remove useless code

guparan committed:
[CMake] FIX SceneCreator and SofaTest build dirs

guparan committed:
[SofaMacros] ADD sofa_set_int and sofa_extlib_find_package

sofa_set_int: define a variable to 0 or 1 to be usable by both config.h.and or XXXConfig.cmake.in

sofa_extlib_find_package: call find_package + call sofa_set_int to create a XXX_HAVE_YYY variable

guparan committed:
[SofaAdvanced] Remove MiniFlowVR (not used)

guparan committed:
[SofaMacros] ADD BOTH_SCOPES option to sofa_set_int and sofa_extlib_find_package

guparan committed:
[SofaMacros] FIX add_subdirectory in sofa_add_generic

"name" has nothing to do as add_subdirectory 2nd argument

guparan committed:
[SofaMacros] Rename sofa_extlib_find_package to sofa_find_package

guparan committed:
[SofaMacros] ADD debug_print_target_properties

Useful to debug.
Prints all properties of a target.

guparan committed:
[plugins] Handle SofaHighOrder first

SofaHighOrderTopology is needed by Flexible

guparan committed:
[libQGLViewer] CLEAN CMake OpenGL link

guparan committed:
[CMake] MOVE configure_file to the end in package CMakeLists

Because some variables defined above may be needed in the configured files

guparan committed:
[SofaSimulationGraph] MOVE test inclusion to modules/tests

guparan committed:
[SofaSimulationGraph] FIX missing config.h

guparan committed:
[SofaSimulationGraph_test] CLEAN CMakeLists.txt

guparan committed:
[SofaSimulationCommon] CLEAN TinyXML dependency

- Add forward declaration to TinyXML classes
- Add find_package
- Link TinyXML privately

guparan committed:
[SofaMacros] Change sofa_set_int to sofa_set_bool

Setting variables as boolean permits to use them with #if or directly with if() without implicit conversion warning.

guparan committed:
[SofaHaptics] Remove unneeded SOFA_HAVE_SOFAHAPTICS

guparan committed:
[SofaComponentMisc] CLEAN CMakeLists.txt

guparan committed:
[All] CLEAN usage of SOFA_DETECTIONOUTPUT_FREEMOTION

Always using the same name from CMake option definition to C++ #ifdef blocks

guparan committed:
[SofaHelper] Remove unneeded SOFA_HAVE_BOOST

The addition of TimeoutWatchdog is already controlled by Boost_THREAD_FOUND in CMakeLists.txt

guparan committed:
[LeapMotion] Remove unneeded boost include

guparan committed:
[SofaOpenglVisual] FIX Config.cmake.in

guparan committed:
[SofaSphFluid] FIX deprecated ExtVec3Types intanciation

guparan committed:
[SofaSphFluid] Remove unneeded SOFA_HAVE_SOFASPHFLUID from Config.cmake.in

guparan committed:
[SofaVolumetricData] Remove unneeded MiniFlowVR link

guparan committed:
[projects] Remove unneeded SOFA_HAVE_QT_LIBRARIES

guparan committed:
[SofaOpenglVisual] Remove unneeded SOFA_HAVE_SOFAOPENGLVISUAL

guparan committed:
[SofaFramework_test] Privately link GTest

guparan committed:
[SofaCore] Remove duplicated addition of AspectPool

guparan committed:
[SofaExplicitOdeSolver] Remove unneeded include

guparan committed:
[SofaMacros] FIX sofa_copy_libraries when called out of SOFA

CMAKE_RUNTIME_OUTPUT_DIRECTORY can be undefined

guparan committed:
[Flexible] Remove outdated SOFA_LINK_CLASS blocks

guparan committed:
[SofaMacros] ADD auto configure of "*.h.in" in sofa_install_targets

/!\ ONLY IF USING sofa_install_targets (basically all plugins and modules) /!\

No more need to configure "*.h.in" files by hand in the CMakeLists.txt.
Instead, simply add those files in HEADER_FILES.

guparan committed:
[QGLViewer] FIX out-of-SOFA use

- use sofa_find_package to set QGLVIEWER_HAVE_* variables
- use those variables to find_package dependencies in Config.cmake.in

guparan committed:
[SofaExporter_test] CLEAN CMakeLists.txt

guparan committed:
[SofaKernel] ADD TinyXML links where it is used

guparan committed:
[SofaSimulationCommon] Support only TinyXML parser (no more libXML)

guparan committed:
[SofaKernel] CLEAN CMake and SOFA_HAVE_* vars

- Move a LOT of things done in SofaFramework/CMakeLists.txt to where they are actually needed.
- Use sofa_find_package to define accurate *_HAVE_* variables.
- Fix dependency finding in *Config.cmake.in
- Change sofa/helper/helper.h into sofa/helper/helper.h.in to propagate SOFAHELPER_HAVE_* variables in sources.

guparan committed:
[SofaMisc] CLEAN CMake and SOFA_HAVE_* vars

guparan committed:
[SofaGeneral] CLEAN CMake and SOFA_HAVE_* vars

guparan committed:
[SofaSparseSolver] CLEAN CMake and SOFA_HAVE_* vars

guparan committed:
[SofaPreconditioner] CLEAN CMake and SOFA_HAVE_* vars

guparan committed:
[SofaOpenglVisual] Complete pluginization + better tests inclusion

guparan committed:
[SofaHaptics] Complete pluginization

guparan committed:
[SofaExporter] CLEAN CMake and SOFA_HAVE_* vars + Complete pluginization + Better tests inclusion

guparan committed:
[tests] Pluginized modules add their own tests

For that, <sofa_root>/modules directory must be added after SofaTest in main CMakeLists.txt

guparan committed:
[SofaBoundaryCondition_test] CLEAN SofaSparseSolver dependency

guparan committed:
[SofaGui] CLEAN CMake and SOFA_HAVE_* vars

- Simplify SofaGui/CMakeLists.txt
- Move all find_package calls where they should be
- Change sofa/gui/qt/SofaGUIQt.h to sofa/gui/qt/SofaGuiQt.h.in

guparan committed:
[meshconv] Remove unneeded SOFA_HAVE_MINIFLOWVR

guparan committed:
[projects] FIX projects using old SOFA_HAVE_BOOST

Replaced by SOFAHELPER_HAVE_BOOST

guparan committed:
[SofaPython] Remove unneeded SOFA_HAVE_PYTHON

guparan committed:
[SofaMiscCollision] CLEAN dependency management

Correctly finding+linking deps and creating consistent variables to be used in sources or by plugin users.

guparan committed:
[SofaImplicitField] CLEAN dependency management

Correctly finding+linking deps and creating consistent variables to be used in sources or by plugin users.

guparan committed:
[SofaDistanceGrid] CLEAN dependency management

Correctly finding+linking deps and creating consistent variables to be used in sources or by plugin users.

guparan committed:
[SofaCUDA] CLEAN MiniFlowVR dep management

Correctly finding and linking privately MiniFlowVR.

guparan committed:
[SceneCreator] CLEAN dependency management

Correctly finding+linking deps and creating consistent variables to be used in sources or by plugin users.

guparan committed:
[Registration] CLEAN dependency management

Correctly finding+linking deps and creating consistent variables to be used in sources or by plugin users.

guparan committed:
[image] CLEAN dependency management

Correctly finding+linking deps and creating consistent variables to be used in sources or by plugin users.

guparan committed:
[Flexible] CLEAN dependency management

Correctly finding+linking deps and creating consistent variables to be used in sources or by plugin users.

guparan committed:
[Compliant] CLEAN dependency management

Correctly finding+linking deps and creating consistent variables to be used in sources or by plugin users.

guparan committed:
[ColladaSceneLoader] CLEAN dependency management

Correctly finding+linking deps and creating consistent variables to be used in sources or by plugin users.

guparan committed:
[CImgPlugin] CLEAN dependency management

Correctly finding+linking deps and creating consistent variables to be used in sources or by plugin users.

guparan committed:
[PluginExample] CLEAN dependency management

Correctly finding+linking deps and creating consistent variables to be used in sources or by plugin users.

guparan committed:
[All] Install ZLIB only on Windows

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into improve_extlibs_integration

# Conflicts:
#	SofaGui/CMakeLists.txt
#	SofaGui/config.h.in
#	SofaKernel/SofaFramework/SofaMacros.cmake
#	SofaKernel/framework/sofa/core/CMakeLists.txt
#	applications/plugins/SceneCreator/CMakeLists.txt
#	applications/plugins/SofaSphFluid/SPHFluidSurfaceMapping.cpp
#	applications/sofa/gui/BaseViewer.cpp
#	applications/sofa/gui/BaseViewer.h
#	applications/sofa/gui/qt/CMakeLists.txt
#	applications/sofa/gui/qt/viewer/SofaViewer.cpp

guparan committed:
[SofaGui] Use SOFA_NO_OPENGL to build an OpenGL-free version of SofaGuiQt

This is a remake of #1121 without creating a new project.

guparan committed:
[SofaGuiMain] CLEAN CMakeLists.txt

guparan committed:
[SofaGuiQt] CLEAN CMakeLists.txt

SofaGui/CMakeLists.txt should not check Qt existence.
If SOFAGUI_QT is ON, then Qt is REQUIRED.

guparan committed:
[SofaGuiQt] CLEAN CMake Qt integration

guparan committed:
[QGLViewer] CLEAN CMake Qt integration

guparan committed:
[SofaGui] FIX SOFA_HAVE_FFMPEG_EXEC renamed

New variable is SOFAHELPER_HAVE_FFMPEG_EXEC

guparan committed:
[image] FIX typo in Config.cmake.in

guparan committed:
[CMake] FIX sofa_set_bool when used with undefined value

guparan committed:
[SofaCUDA] Revert to plain signature for target_link_libraries

Due to this error (tested with CMake 3.5 and CMake 3.13.3):
CMake Error at applications/plugins/SofaCUDA/CMakeLists.txt:317 (target_link_libraries):
  The plain signature for target_link_libraries has already been used with
  the target "SofaCUDA".  All uses of target_link_libraries with a target
  must be either all-keyword or all-plain.

  The uses of the plain signature are here:

   * /usr/share/cmake-3.5/Modules/FindCUDA.cmake:1660 (target_link_libraries)

guparan committed:
[SofaGuiQt] FIX file not found in GUI.ui

fredroy commented on SofaGui/SofaGuiConfig.cmake.in:
```diff
@@ -2,31 +2,38 @@
+        find_package(QGLViewer QUIET REQUIRED)
+    endif()
+    if(SOFAGUIQT_HAVE_QWT)
+        find_package(QGLViewer QUIET REQUIRED)
```
Typo ?  find_package(qwt QUIET REQUIRED) instead ?

guparan replied:
Thanks! :+1:

guparan committed:
[SofaMacros] Propagate sofa_set_bool fix to sofa_find_package

guparan committed:
[SofaGui] FIX typo in Qwt finding + improve

guparan committed:
[SofaCUDA] FIX missing config.h

guparan committed:
[CGALPlugin] FIX missing config.h

guparan committed:
[SofaCUDA] FIX includes order

guparan committed:
[SofaGui] FIX #ifdef to #if for SOFAHELPER_HAVE_FFMPEG_EXEC

SOFAHELPER_HAVE_FFMPEG_EXEC is a boolean

guparan committed:
[Modeler] FIX TinyXML dependency

guparan committed:
[SofaMiscFem_test] FIX missing test

guparan committed:
[SofaSimulationGraph_test] FIX missing test

guparan committed:
[CMake] Change order to have SofaSparseSolver when building tests

guparan commented:
[ci-build][with-all-tests]

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into improve_extlibs_integration

# Conflicts:
#	applications/sofa/gui/qt/CMakeLists.txt
#	applications/sofa/gui/qt/gl/CMakeLists.txt
#	applications/sofa/gui/qt/viewer/GLBackend.cpp
#	extlibs/CMakeLists.txt

guparan committed:
[All] Use cmakedefine01 to ensure correct value

even if variable goes from TRUE to UNDEFINED (old way let the #define with no value, leading to errors)

guparan committed:
[CMake] RENAME sofa_set_bool to sofa_set_01

Better consistency with #cmakedefine01

epernod committed:
[cmake] FIX cmake configuration when SOFA_TEST is disabled

epernod committed:
[SofaExporter] FIX compilation when sofa_python is not activated.

guparan commented on modules/SofaExporter/src/SofaExporter/initExporter.cpp:
```diff
@@ -60,11 +60,12 @@ void initExternalModule()
     {
         first = false;
-        if(SOFAEXPORTER_HAVE_SOFAPYTHON)
+#if SOFAEXPORTER_HAVE_SOFAPYTHON
```
Thanks, this was indeed broken.

guparan commented on modules/SofaExporter/CMakeLists.txt:
```diff
@@ -81,6 +81,13 @@ sofa_generate_package(
+    option(SOFAEXPORTER_BUILD_TESTS "Compile the automatic tests" ON)
+else()
+    set(SOFAEXPORTER_BUILD_TESTS FALSE)
+endif()
```
I don't agree with this approach. 
SOFA_BUILD_TEST is for SOFA tests. SofaExporter is now modularized so it should be considered as out-of-SOFA.

My view is:
- SOFA old modules (that are not real modules) do not integrate their tests themselves because of SofaTest dependencies (see the order of `add_subdirectory` calls in main CMakeLists)
- new modules ("modularized") and plugins and projects must be buildable out of SOFA so they integrate their tests themselves and do not depend on a SOFA internal variable. Those tests, though, can have a SofaTest dependency (SofaExporter_test does) and thus may fail to generate if SofaTest is not found.

In this case for example, I prefer an explicit SOFAEXPORTER_BUILD_TESTS that may lead to a failure because SofaExporter_test couldn't find SofaTest than a magic check above SofaExporter_test setting the variable OFF beforehand.

What could be changed though is the way SofaExporter_test finds SofaTest
```cmake
find_package(SofaTest QUIET)
if(NOT SofaTest_FOUND)
    message(SEND_ERROR "Couldn't find SofaTest. If building SOFA, you need to activate SOFA_BUILD_TESTS")
endif()
```

epernod replied:
ok I understand that tests from module should be independant but they need SofaTest lib.
In my case I don't compile SofaTest (bad I know).  But unchecking SofaTest failed at cmake configuration because SOFAEXPORTER_BUILD_TESTS need SofaTest. So it means I would need to uncheck each TEST in the cmake configuration...

Maybe the good approach is (not sure it works): 
```cmake
if (SOFA_BUILD_TESTS)
    option(SOFAEXPORTER_BUILD_TESTS "Compile the automatic tests" ON)
else()
    option(SOFAEXPORTER_BUILD_TESTS "Compile the automatic tests" OFF)
endif()
```

and in the SOFAExporter_test cmake:

```cmake
find_package(SofaTest REQUIRED)
if(NOT SofaTest_FOUND)
    message(SEND_ERROR "Couldn't find SofaTest. If building SOFA, you need to activate SOFA_BUILD_TESTS")
endif()
```

guparan replied:
I think the problem behind is that SOFA_BUILD_TESTS should not be used to check SofaTest_FOUND.
SOFA_BUILD_TESTS should be only used to ON/OFF the old module tests (all at once).

Being able to ON/OFF specific tests is a good thing in my opinion.
Here is a compromise, in SofaExporter_test CMakeLists:
```cmake
find_package(SofaTest QUIET)
if(NOT SofaTest_FOUND)
    message(STATUS "Couldn't find SofaTest. ${PROJECT_NAME} will not be built.")
    return()
endif()
```

This is more or less the same idea but keeping the check inside of SofaExporter_test.

epernod committed:
[cmake] update condition to activate specific package test. Use find_package(SofaTest)

epernod reviewed (approved):
Not sure it works fully as expected but for me cmake generation works as well as compilation and make install.
I was also able to create a project using SOFA as lib. It is not totally easy going but it works.
